includes/compat/trait-wc-stripe-subscriptions.php

```diff
@@ -244,6 +246,103 @@ public function process_change_subscription_payment_method( $order_id ) {
 		}
 	}
 
+	/**
+	 * Process the payment method change with deferred intent.
+	 *
+	 * @param int $subscription_id
+	 *
+	 * @return array
+	 */
+	public function process_change_subscription_payment_with_deferred_intent( $subscription_id ) {
+		$subscription = wcs_get_subscription( $subscription_id );
+
+		if ( ! $subscription ) {
+			return [
+				'result'   => 'failure',
+				'redirect' => '',
+			];
+		}
+
+		try {
+			$payment_information = $this->prepare_payment_information_from_request( $subscription );
+
+			$this->validate_selected_payment_method_type( $payment_information, $subscription->get_billing_country() );
+
+			$payment_method_id     = $payment_information['payment_method'];
+			$selected_payment_type = $payment_information['selected_payment_type'];
+			$upe_payment_method    = $this->payment_methods[ $selected_payment_type ] ?? null;
+
+			// Retrieve the payment method object from Stripe.
+			$payment_method = $this->stripe_request( 'payment_methods/' . $payment_method_id );
+
+			// Throw an exception when the payment method is a prepaid card and it's disallowed.
+			$this->maybe_disallow_prepaid_card( $payment_method );
+
+			// Create a setup intent, or update an existing one associated with the order.
+			$payment_intent = $this->process_setup_intent_for_order( $subscription, $payment_information );
+
+			// Handle saving the payment method in the store.
+			if ( $payment_information['save_payment_method_to_store'] && $upe_payment_method && $upe_payment_method->get_id() === $upe_payment_method->get_retrievable_type() ) {
+				$this->handle_saving_payment_method(
+					$subscription,
+					$payment_information['payment_method_details'],
+					$selected_payment_type
+				);
+			}
+
+			$redirect           = $this->get_return_url( $subscription );
```

Comment:
Do you know why we use this method to get the redirect URL? This returns the my account page `https://example/my-account/`. Shouldn't it be the subscription page.

Reply:
> Shouldn't it be the subscription page.

Yeah so inside WooCommerce Subscriptions, the `WC_Subscriptions_Change_Payment_Gateway` class filters this `$this->get_return_url()` to return the `my-account` URL.

On successful update, it then changes this redirect URL to be the view subscriptions page (see [this code](https://github.com/Automattic/woocommerce-subscriptions-core/blob/6bfbf519df799dde8cd9f39e6c93e820e5170352/includes/class-wc-subscriptions-change-payment-gateway.php#L341-L343))

Inside our 3DS card handler, we are also manually setting the return URL to be `$subscription->get_view_order_url()` on success.

In your testing, did you get redirected incorrectly?

Reply:
I was being redirected to the my account page after changing the subscription to Cash App but I think that's fixed now by making sure we use the same flow that 3DS does to confirm the change payment flow after completing the Cash App setup.